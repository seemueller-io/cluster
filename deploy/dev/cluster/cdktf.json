{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "4acdc42d-3176-4b8e-a8ab-50a2f1152e55",
  "sendCrashReports": "false",
  "terraformProviders": [
    "kreuzwerker/docker@~> 3.0",
    "hashicorp/kubernetes@~> 2.0",
    "hashicorp/helm@~> 2.0",
    "hashicorp/null@~> 3.0"
  ],
  "terraformModules": [],
  "context": {

  }
}
