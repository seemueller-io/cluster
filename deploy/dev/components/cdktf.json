{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "92d0a2d7-b985-4c11-b887-5f629f3aa198",
  "sendCrashReports": "false",
  "terraformProviders": [
    "helm@~> 2.0",
    "kubernetes@~> 2.0",
    "null@~> 3.0"
  ],
  "terraformModules": [],
  "context": {

  }
}
