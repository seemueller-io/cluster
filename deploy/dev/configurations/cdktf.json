{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "zitadel-configurations",
  "sendCrashReports": "false",
  "terraformProviders": [
    "zitadel/zitadel@~> 1.0"
  ],
  "terraformModules": [],
  "context": {

  }
}