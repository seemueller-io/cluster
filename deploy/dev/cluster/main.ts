import {Construct} from "constructs";
import {App, TerraformStack} from "cdktf";
import {DockerProvider} from "./.gen/providers/docker/provider";
import {Container} from "./.gen/providers/docker/container";
import {Image} from "./.gen/providers/docker/image";
import {KubernetesProvider} from "./.gen/providers/kubernetes/provider";
import {ConfigMapV1} from "./.gen/providers/kubernetes/config-map-v1";
import {NullProvider} from "./.gen/providers/null/provider";
import {Resource} from "./.gen/providers/null/resource";

export class DockerRegistryStack extends TerraformStack {
    public readonly registryContainer: Container;
    public readonly regName: string = "kind-registry";
    public readonly regPort: string = "5001";

    constructor(scope: Construct, id: string) {
        super(scope, id);

        // Configure providers
        new DockerProvider(this, "docker", {});

        // 1. Create registry container (equivalent to the first part of create-cluster.sh)
        // Pull the registry image
        const registryImage = new Image(this, "registry-image", {
            name: "registry:2",
            keepLocally: true,
        });

        // Create the registry container
        this.registryContainer = new Container(this, "kind-registry", {
            name: this.regName,
            image: registryImage.imageId,
            ports: [{
                internal: 5000,
                external: parseInt(this.regPort),
                ip: "127.0.0.1",
            }],
            restart: "always",
            networksAdvanced: [{
                name: "bridge",
            }],
        });
    }
}


const kindClusterConfig = `kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry]
    config_path = "/etc/containerd/certs.d"
nodes:
- role: control-plane
  extraPortMappings:
  - containerPort: 30080
    hostPort: 80
    protocol: TCP
  - containerPort: 30443
    hostPort: 443
    protocol: TCP
`


export class KindClusterStack extends TerraformStack {
    public readonly networkConnection: Resource;
    public readonly kindCluster: Resource;
    public readonly registryConfig: Resource;

    constructor(scope: Construct, id: string, registryStack: DockerRegistryStack) {
        super(scope, id);

        // Add dependency on the registry stack
        this.addDependency(registryStack);

        // Configure providers
        new NullProvider(this, "null", {});

        // 2. Create Kind cluster with configuration
        // This uses a null resource to execute the kind create cluster command
        this.kindCluster = new Resource(this, "kind-cluster", {
            triggers: {
                config: kindClusterConfig
            },
            provisioners: [
                {
                    type: "local-exec",
                    command: `echo '${kindClusterConfig}' | kind create cluster --config=-`,
                    when: 'create'
                },
                {
                    type: "local-exec",
                    command: `kind delete cluster -n kind`,
                    when: 'destroy'
                }
            ],
        });

        // 3. Configure registry for cluster nodes
        this.registryConfig = new Resource(this, "registry-config", {
            provisioners: [
                {
                    type: "local-exec",
                    command: `
            REGISTRY_DIR="/etc/containerd/certs.d/localhost:${registryStack.regPort}"
            for node in $(kind get nodes); do
              docker exec "$node" mkdir -p "$REGISTRY_DIR"
              echo '[host."http://${registryStack.regName}:5000"]' | docker exec -i "$node" cp /dev/stdin "$REGISTRY_DIR/hosts.toml"
            done
          `,
                },
            ],
            dependsOn: [this.kindCluster],
        });

        // 4. Connect registry to cluster network
        this.networkConnection = new Resource(this, "network-connection", {
            provisioners: [
                {
                    type: "local-exec",
                    command: `
            if [ "$(docker inspect -f='{{json .NetworkSettings.Networks.kind}}' "${registryStack.regName}")" = 'null' ]; then
              docker network connect "kind" "${registryStack.regName}"
            fi
          `,
                },
            ],
            dependsOn: [this.registryConfig],
        });
    }
}


export class ClusterConfigStack extends TerraformStack {
    constructor(scope: Construct, id: string, registryStack: DockerRegistryStack, kindClusterStack: KindClusterStack) {
        super(scope, id);

        // Add dependency on the kind cluster stack
        this.addDependency(kindClusterStack);

        // Configure Kubernetes provider after cluster is created
        new KubernetesProvider(this, "kubernetes", {
            configPath: "~/.kube/config",
            configContext: "kind-kind",
        });

        // Create Kubernetes ConfigMap to document the local registry
        new ConfigMapV1(this, "local-registry-hosting", {
            metadata: {
                name: "local-registry-hosting",
                namespace: "kube-public",
            },
            data: {
                "localRegistryHosting.v1": `host: "localhost:${registryStack.regPort}"
help: "https://kind.sigs.k8s.io/docs/user/local-registry/"`,
            },
        });
    }
}

const app = new App();
const registryStack = new DockerRegistryStack(app, "docker-registry");
const kindClusterStack = new KindClusterStack(app, "kind-cluster", registryStack);

new ClusterConfigStack(app, "cluster-config", registryStack, kindClusterStack);
app.synth();
