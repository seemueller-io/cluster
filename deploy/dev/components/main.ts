import { Construct } from "constructs";
import { App, TerraformStack, TerraformOutput } from "cdktf";
import { HelmProvider } from "./.gen/providers/helm/provider";
import { Release } from "./.gen/providers/helm/release";
import { KubernetesProvider } from "./.gen/providers/kubernetes/provider";
import { NullProvider } from "./.gen/providers/null/provider";
import { Resource } from "./.gen/providers/null/resource";

export class ClusterComponentsStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    // Configure providers
    new HelmProvider(this, "helm", {
      kubernetes: {
        configPath: "~/.kube/config",
        configContext: "kind-kind",
      },
    });
    new KubernetesProvider(this, "kubernetes", {
      configPath: "~/.kube/config",
      configContext: "kind-kind",
    });
    new NullProvider(this, "null", {});

    // 1. Install cert-manager for TLS certificate management
    const certManager = new Release(this, "cert-manager", {
      name: "cert-manager",
      repository: "oci://quay.io/jetstack/charts",
      chart: "cert-manager",
      version: "v1.18.2",
      namespace: "cert-manager",
      createNamespace: true,
      wait: true,
      set: [
        {
          name: "crds.enabled",
          value: "true",
        },
      ],
    });

    // 2. Install Traefik ingress controller
    const traefik = new Release(this, "traefik", {
      name: "traefik",
      repository: "https://traefik.github.io/charts",
      chart: "traefik",
      version: "36.3.0",
      namespace: "ingress",
      createNamespace: true,
      wait: true,
      values: [
        `logs:
  general:
    level: DEBUG
additionalArguments:
  - "--serverstransport.insecureskipverify=true"
service:
  type: NodePort
ports:
  web:
    nodePort: 30080
    redirections:
      entryPoint:
        to: websecure
        scheme: https
        permanent: true
  websecure:
    nodePort: 30443
ingressClass:
  enabled: true
  isDefaultClass: true`,
      ],
      dependsOn: [certManager],
    });

    // 3. Install PostgreSQL database
    const postgresql = new Release(this, "postgresql", {
      name: "db",
      repository: "https://charts.bitnami.com/bitnami",
      chart: "postgresql",
      version: "12.10.0",
      namespace: "default",
      wait: true,
      values: [
        `primary:
  pgHbaConfiguration: |
    host all all all trust`,
      ],
      dependsOn: [traefik],
    });

    // 4. Install Zitadel
    const zitadel = new Release(this, "zitadel", {
      name: "my-zitadel",
      repository: "https://charts.zitadel.com",
      chart: "zitadel",
      namespace: "default",
      wait: true,
      values: [
        `zitadel:
  masterkey: x123456789012345678901234567891y
  configmapConfig:
    Log:
      Level: debug
    ExternalDomain: machine.127.0.0.1.sslip.io
    ExternalPort: 443
    TLS:
      Enabled: false
    FirstInstance:
      Org:
        Machine:
          Machine:
            Username: zitadel-admin-sa
            Name: Admin
          MachineKey:
            ExpirationDate: "2026-01-01T00:00:00Z"
            Type: 1
        # PAT:
        #   ExpirationDate: "2026-01-01T00:00:00Z"
    Database:
      Postgres:
        Host: db-postgresql
        Port: 5432
        Database: zitadel
        MaxOpenConns: 20
        MaxIdleConns: 10
        MaxConnLifetime: 30m
        MaxConnIdleTime: 5m
        User:
          Username: postgres
          SSL:
            Mode: disable
        Admin:
          Username: postgres
          SSL:
            Mode: disable
ingress:
  enabled: true
login:
  ingress:
    enabled: true`,
      ],
      dependsOn: [postgresql],
    });

    // 5. Wait for cert-manager CRDs to be available
    const waitForCertManagerCRDs = new Resource(this, "wait-for-cert-manager-crds", {
      triggers: {
        cert_manager_dependency: certManager.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: "kubectl wait --for=condition=established --timeout=120s crd/clusterissuers.cert-manager.io || kubectl get crd clusterissuers.cert-manager.io",
          when: "create",
        },
      ],
      dependsOn: [certManager],
    });

    // 6. Create self-signed certificate issuer and certificate using kubectl apply
    const createTLSResources = new Resource(this, "create-tls-resources", {
      triggers: {
        crd_dependency: waitForCertManagerCRDs.id,
        zitadel_dependency: zitadel.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `cat <<EOF | kubectl apply -f -
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: selfsigned-issuer
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: zitadel-cert
  namespace: default
spec:
  secretName: zitadel-tls
  issuerRef:
    name: selfsigned-issuer
    kind: ClusterIssuer
  commonName: machine.127.0.0.1.sslip.io
  dnsNames:
  - machine.127.0.0.1.sslip.io
EOF`,
          when: "create",
        },
      ],
      dependsOn: [waitForCertManagerCRDs, zitadel],
    });

    // 7. Wait for certificate to be ready
    const waitForCertificate = new Resource(this, "wait-for-certificate", {
      triggers: {
        tls_resources_dependency: createTLSResources.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: "kubectl wait --for=condition=ready certificate zitadel-cert -n default --timeout=120s || true",
          when: "create",
        },
      ],
      dependsOn: [createTLSResources],
    });

    // 8. Patch ingresses with TLS configuration
    const patchIngresses = new Resource(this, "patch-ingresses", {
      triggers: {
        wait_dependency: waitForCertificate.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `
            kubectl patch ingress my-zitadel -n default --type='merge' -p='{"spec":{"tls":[{"hosts":["machine.127.0.0.1.sslip.io"],"secretName":"zitadel-tls"}]}}' || true
            kubectl patch ingress my-zitadel-login -n default --type='merge' -p='{"spec":{"tls":[{"hosts":["machine.127.0.0.1.sslip.io"],"secretName":"zitadel-tls"}]}}' || true
          `,
          when: "create",
        },
      ],
      dependsOn: [waitForCertificate],
    });

    // 9. Configure custom SSL and extract certificate
    const configureSSL = new Resource(this, "configure-ssl", {
      triggers: {
        patch_dependency: patchIngresses.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `
            # Extract certificate and add to system trust store
            kubectl get secret zitadel-tls -n default -o jsonpath='{.data.tls\\.crt}' | base64 -d > ./certs/zitadel-cert.crt || true
          `,
          when: "create",
        },
      ],
      dependsOn: [patchIngresses],
    });

    // 10. Wait for Zitadel service account secret and extract credentials
    const extractCredentials = new Resource(this, "extract-credentials", {
      triggers: {
        ssl_dependency: configureSSL.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `echo 'Credential extraction would run during apply'`,
          when: "create",
        },
      ],
      dependsOn: [configureSSL],
    });

    // 11. Verify Zitadel accessibility
    const verifyZitadel = new Resource(this, "verify-zitadel", {
      triggers: {
        credentials_dependency: extractCredentials.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `echo 'Zitadel verification would run during apply'`,
          when: "create",
        },
      ],
      dependsOn: [extractCredentials],
    });

    // 12. Output success message
    new Resource(this, "completion-message", {
      triggers: {
        verification_dependency: verifyZitadel.id,
      },
      provisioners: [
        {
          type: "local-exec",
          command: `echo 'Installation completed successfully!'`,
          when: "create",
        },
      ],
      dependsOn: [verifyZitadel],
    });

    // Output important information
    new TerraformOutput(this, "zitadel_url", {
      value: "https://machine.127.0.0.1.sslip.io/ui/console?login_hint=zitadel-admin@zitadel.machine.127.0.0.1.sslip.io",
      description: "Zitadel Console URL",
    });

    new TerraformOutput(this, "admin_credentials", {
      value: "zitadel-admin@zitadel.machine.127.0.0.1.sslip.io / Password1!",
      description: "Default admin credentials",
    });
  }
}

const app = new App();
new ClusterComponentsStack(app, "cluster-components");
app.synth();
