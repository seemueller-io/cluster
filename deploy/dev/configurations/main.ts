import {Construct} from "constructs";
import {App, TerraformOutput, TerraformStack} from "cdktf";

import {Org} from "./.gen/providers/zitadel/org";
import {Project} from "./.gen/providers/zitadel/project";
import {ApplicationOidc} from "./.gen/providers/zitadel/application-oidc";
import {HumanUser} from "./.gen/providers/zitadel/human-user";
import {ZitadelProvider} from "./.gen/providers/zitadel/provider";

import * as path from "node:path";
import {readFileSync} from "fs";

export class ZitadelStack extends TerraformStack {
    public readonly createdOrg: Org;
    public readonly createdProject: Project;
    public readonly createdApp: ApplicationOidc;
    public readonly createdUser: HumanUser;

    constructor(scope: Construct, id: string) {
        super(scope, id);

        const provider = new ZitadelProvider(this, "zitadel", {
            domain: "machine.127.0.0.1.sslip.io",   // your instance URL
            jwtProfileJson: JSON.stringify(JSON.parse(readFileSync(path.resolve("zitadel-admin-sa.json").toString(), 'utf-8'))),
        });


        this.createdOrg = new Org(this, "org", {
            name: "makers",
            provider: provider,
        });

        this.createdProject = new Project(this, "project", {
            name: "makers-project",
            orgId: this.createdOrg.id,
            provider: provider,
        });

        this.createdApp = new ApplicationOidc(this, "app", {
            name: "makers-app",
            projectId: this.createdProject.id,
            orgId: this.createdOrg.id,
            grantTypes: ["OIDC_GRANT_TYPE_AUTHORIZATION_CODE"],
            redirectUris: ["http://localhost:3000/callback"],
            responseTypes: ["OIDC_RESPONSE_TYPE_CODE"],
            provider: provider,
            dependsOn: [this.createdProject],
        });

        this.createdUser = new HumanUser(this, "user", {
            userName: "makers-user",
            email: "makers-user@example.com",
            firstName: "Makers",
            lastName: "User",
            displayName: "Makers User",
            orgId: this.createdOrg.id,
            initialPassword: "TempPassword123!",
            isEmailVerified: true,
            provider: provider,
        });

        new TerraformOutput(this, "client_id", {
            value: this.createdApp.clientId,
            description: "The client ID of the OIDC application",
            sensitive: true,
        });

        new TerraformOutput(this, "client_secret", {
            value: this.createdApp.clientSecret,
            description: "The client secret of the OIDC application",
            sensitive: true,
        });

        new TerraformOutput(this, "user_login_names", {
            value: this.createdUser.loginNames,
            description: "The login names of the created user",
            sensitive: true,
        });

        new TerraformOutput(this, "user_password", {
            value: this.createdUser.initialPassword,
            description: "The password of the created user",
            sensitive: true,
        });

        new TerraformOutput(this, "user_preferred_login_name", {
            value: this.createdUser.preferredLoginName,
            description: "The preferred login name of the created user",
            sensitive: true,
        });

        new TerraformOutput(this, "user_state", {
            value: this.createdUser.state,
            description: "The state of the created user",
            sensitive: true,
        });

        new TerraformOutput(this, "created_org", {
            value: {
                id: this.createdOrg.id
            },
            description: "The client ID of the OIDC application",
            sensitive: true,
        });

        new TerraformOutput(this, "created_project", {
            value: {
                id: this.createdProject.id,
                name: this.createdProject.name,
            },
            description: "The client ID of the OIDC application",
            sensitive: true,
        });

    }
}

const app = new App();

new ZitadelStack(app, "zitadel-dev");

app.synth();






