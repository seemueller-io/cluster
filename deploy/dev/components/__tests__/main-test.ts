// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import "cdktf/lib/testing/adapters/jest"; // Load types for expect matchers
import { Testing } from "cdktf";
import { ClusterComponentsStack } from "../main";

describe("ClusterComponentsStack", () => {
  describe("Resource Creation", () => {
    it("should create all required Helm releases", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      // Check for all Helm releases
      expect(synthesized).toContain("helm_release");
      expect(synthesized).toContain('"name": "cert-manager"');
      expect(synthesized).toContain('"name": "traefik"');
      expect(synthesized).toContain('"name": "db"');
      expect(synthesized).toContain('"name": "my-zitadel"');
    });

    it("should create cert-manager with correct configuration", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"chart": "cert-manager"');
      expect(synthesized).toContain('"version": "v1.18.2"');
      expect(synthesized).toContain('"namespace": "cert-manager"');
      expect(synthesized).toContain('"create_namespace": true');
      expect(synthesized).toContain('"wait": true');
      expect(synthesized).toContain('"name": "crds.enabled"');
      expect(synthesized).toContain('"value": "true"');
    });

    it("should create Traefik with correct configuration", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"chart": "traefik"');
      expect(synthesized).toContain('"version": "36.3.0"');
      expect(synthesized).toContain('"namespace": "ingress"');
      expect(synthesized).toContain("https://traefik.github.io/charts");
      expect(synthesized).toContain("traefik-values.yaml");
    });

    it("should create PostgreSQL with correct configuration", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"chart": "postgresql"');
      expect(synthesized).toContain('"version": "12.10.0"');
      expect(synthesized).toContain("charts.bitnami.com/bitnami");
      expect(synthesized).toContain("postgres-values.yaml");
    });

    it("should create Zitadel with correct configuration", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"chart": "zitadel"');
      expect(synthesized).toContain("charts.zitadel.com");
      expect(synthesized).toContain("zitadel-values.yaml");
    });

    it("should create TLS resources using kubectl apply", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      // Check for null_resource with kubectl apply command containing TLS resources
      expect(synthesized).toContain("null_resource");
      expect(synthesized).toContain("create-tls-resources");
      expect(synthesized).toContain("kubectl apply -f -");
      expect(synthesized).toContain("kind: ClusterIssuer");
      expect(synthesized).toContain("name: selfsigned-issuer");
      
      // Check for Certificate in the kubectl apply command
      expect(synthesized).toContain("kind: Certificate");
      expect(synthesized).toContain("name: zitadel-cert");
      expect(synthesized).toContain("secretName: zitadel-tls");
      expect(synthesized).toContain("machine.127.0.0.1.sslip.io");
    });

    it("should create null resources for operations", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain("null_resource");
      expect(synthesized).toContain("kubectl wait --for=condition=ready certificate");
      expect(synthesized).toContain("kubectl patch ingress");
      expect(synthesized).toContain("kubectl get secret zitadel-tls");
    });
  });

  describe("Resource Dependencies", () => {
    it("should have proper resource dependencies", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      // Verify that resources have dependencies
      expect(synthesized).toContain("depends_on");
    });

    it("should ensure correct deployment order", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      // Traefik should depend on cert-manager
      const traefikSection = synthesized.match(/"traefik"[\s\S]*?"depends_on"[\s\S]*?"helm_release\.cert-manager"/);
      expect(traefikSection).toBeTruthy();
      
      // PostgreSQL should depend on Traefik
      const postgresSection = synthesized.match(/"postgresql"[\s\S]*?"depends_on"[\s\S]*?"helm_release\.traefik"/);
      expect(postgresSection).toBeTruthy();
      
      // Zitadel should depend on PostgreSQL
      const zitadelSection = synthesized.match(/"zitadel"[\s\S]*?"depends_on"[\s\S]*?"helm_release\.postgresql"/);
      expect(zitadelSection).toBeTruthy();
    });
  });

  describe("Providers Configuration", () => {
    it("should configure all required providers", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"provider"');
      expect(synthesized).toContain('"helm"');
      expect(synthesized).toContain('"kubernetes"');
      expect(synthesized).toContain('"null"');
    });

    it("should configure Kubernetes provider with correct context", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"config_path": "~/.kube/config"');
      expect(synthesized).toContain('"config_context": "kind-kind"');
    });
  });

  describe("Outputs", () => {
    it("should create Terraform outputs for important information", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain('"output"');
      expect(synthesized).toContain('"zitadel_url"');
      expect(synthesized).toContain('"admin_credentials"');
      expect(synthesized).toContain("https://machine.127.0.0.1.sslip.io/ui/console");
      expect(synthesized).toContain("zitadel-admin@zitadel.machine.127.0.0.1.sslip.io");
    });
  });

  describe("TLS Configuration", () => {
    it("should create self-signed certificate issuer", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain("apiVersion: cert-manager.io/v1");
      expect(synthesized).toContain("kind: ClusterIssuer");
      expect(synthesized).toContain("selfSigned: {}");
    });

    it("should create certificate with correct DNS names", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      const synthesized = Testing.synth(stack);
      
      expect(synthesized).toContain("commonName: machine.127.0.0.1.sslip.io");
      expect(synthesized).toContain("dnsNames:");
      expect(synthesized).toContain("- machine.127.0.0.1.sslip.io");
      expect(synthesized).toContain("issuerRef:");
      expect(synthesized).toContain("kind: ClusterIssuer");
    });
  });

  describe("Terraform Configuration Validity", () => {
    it("should generate valid Terraform configuration", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      expect(Testing.fullSynth(stack)).toBeValidTerraform();
    });
  });

  describe("Snapshot Tests", () => {
    it("should match the expected Terraform configuration snapshot", () => {
      const app = Testing.app();
      const stack = new ClusterComponentsStack(app, "test-stack");
      expect(Testing.synth(stack)).toMatchSnapshot();
    });
  });
});
