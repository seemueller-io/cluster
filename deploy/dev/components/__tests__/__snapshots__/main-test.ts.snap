// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ClusterComponentsStack Snapshot Tests should match the expected Terraform configuration snapshot 1`] = `
"{
  "output": {
    "admin_credentials": {
      "description": "Default admin credentials",
      "value": "zitadel-admin@zitadel.machine.127.0.0.1.sslip.io / Password1!"
    },
    "zitadel_url": {
      "description": "Zitadel Console URL",
      "value": "https://machine.127.0.0.1.sslip.io/ui/console?login_hint=zitadel-admin@zitadel.machine.127.0.0.1.sslip.io"
    }
  },
  "provider": {
    "helm": [
      {
        "kubernetes": {
          "config_context": "kind-kind",
          "config_path": "~/.kube/config"
        }
      }
    ],
    "kubernetes": [
      {
        "config_context": "kind-kind",
        "config_path": "~/.kube/config"
      }
    ],
    "null": [
      {
      }
    ]
  },
  "resource": {
    "helm_release": {
      "cert-manager": {
        "chart": "cert-manager",
        "create_namespace": true,
        "name": "cert-manager",
        "namespace": "cert-manager",
        "repository": "oci://quay.io/jetstack/charts",
        "set": [
          {
            "name": "crds.enabled",
            "value": "true"
          }
        ],
        "version": "v1.18.2",
        "wait": true
      },
      "postgresql": {
        "chart": "postgresql",
        "depends_on": [
          "helm_release.traefik"
        ],
        "name": "db",
        "namespace": "default",
        "repository": "https://charts.bitnami.com/bitnami",
        "values": [
          "https://raw.githubusercontent.com/zitadel/zitadel-charts/main/examples/4-machine-user/postgres-values.yaml"
        ],
        "version": "12.10.0",
        "wait": true
      },
      "traefik": {
        "chart": "traefik",
        "create_namespace": true,
        "depends_on": [
          "helm_release.cert-manager"
        ],
        "name": "traefik",
        "namespace": "ingress",
        "repository": "https://traefik.github.io/charts",
        "values": [
          "https://raw.githubusercontent.com/zitadel/zitadel-charts/main/examples/99-kind-with-traefik/traefik-values.yaml"
        ],
        "version": "36.3.0",
        "wait": true
      },
      "zitadel": {
        "chart": "zitadel",
        "depends_on": [
          "helm_release.postgresql"
        ],
        "name": "my-zitadel",
        "namespace": "default",
        "repository": "https://charts.zitadel.com",
        "values": [
          "https://raw.githubusercontent.com/zitadel/zitadel-charts/main/examples/4-machine-user/zitadel-values.yaml"
        ],
        "wait": true
      }
    },
    "null_resource": {
      "completion-message": {
        "depends_on": [
          "null_resource.verify-zitadel"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "echo 'Installation completed successfully!'",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "verification_dependency": "\${null_resource.verify-zitadel.id}"
        }
      },
      "configure-ssl": {
        "depends_on": [
          "null_resource.patch-ingresses"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "\\n            # Extract certificate and add to system trust store\\n            kubectl get secret zitadel-tls -n default -o jsonpath='{.data.tls\\\\.crt}' | base64 -d > /tmp/zitadel-cert.crt || true\\n            sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain /tmp/zitadel-cert.crt || true\\n          ",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "patch_dependency": "\${null_resource.patch-ingresses.id}"
        }
      },
      "create-tls-resources": {
        "depends_on": [
          "null_resource.wait-for-cert-manager-crds",
          "helm_release.zitadel"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "cat <<EOF | kubectl apply -f -\\napiVersion: cert-manager.io/v1\\nkind: ClusterIssuer\\nmetadata:\\n  name: selfsigned-issuer\\nspec:\\n  selfSigned: {}\\n---\\napiVersion: cert-manager.io/v1\\nkind: Certificate\\nmetadata:\\n  name: zitadel-cert\\n  namespace: default\\nspec:\\n  secretName: zitadel-tls\\n  issuerRef:\\n    name: selfsigned-issuer\\n    kind: ClusterIssuer\\n  commonName: machine.127.0.0.1.sslip.io\\n  dnsNames:\\n  - machine.127.0.0.1.sslip.io\\nEOF",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "crd_dependency": "\${null_resource.wait-for-cert-manager-crds.id}",
          "zitadel_dependency": "\${helm_release.zitadel.id}"
        }
      },
      "extract-credentials": {
        "depends_on": [
          "null_resource.configure-ssl"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "echo 'Credential extraction would run during apply'",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "ssl_dependency": "\${null_resource.configure-ssl.id}"
        }
      },
      "patch-ingresses": {
        "depends_on": [
          "null_resource.wait-for-certificate"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "\\n            kubectl patch ingress my-zitadel -n default --type='merge' -p='{\\"spec\\":{\\"tls\\":[{\\"hosts\\":[\\"machine.127.0.0.1.sslip.io\\"],\\"secretName\\":\\"zitadel-tls\\"}]}}' || true\\n            kubectl patch ingress my-zitadel-login -n default --type='merge' -p='{\\"spec\\":{\\"tls\\":[{\\"hosts\\":[\\"machine.127.0.0.1.sslip.io\\"],\\"secretName\\":\\"zitadel-tls\\"}]}}' || true\\n          ",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "wait_dependency": "\${null_resource.wait-for-certificate.id}"
        }
      },
      "verify-zitadel": {
        "depends_on": [
          "null_resource.extract-credentials"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "echo 'Zitadel verification would run during apply'",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "credentials_dependency": "\${null_resource.extract-credentials.id}"
        }
      },
      "wait-for-cert-manager-crds": {
        "depends_on": [
          "helm_release.cert-manager"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "kubectl wait --for=condition=established --timeout=120s crd/clusterissuers.cert-manager.io || kubectl get crd clusterissuers.cert-manager.io",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "cert_manager_dependency": "\${helm_release.cert-manager.id}"
        }
      },
      "wait-for-certificate": {
        "depends_on": [
          "null_resource.create-tls-resources"
        ],
        "provisioner": [
          {
            "local-exec": {
              "command": "kubectl wait --for=condition=ready certificate zitadel-cert -n default --timeout=120s || true",
              "when": "create"
            }
          }
        ],
        "triggers": {
          "tls_resources_dependency": "\${null_resource.create-tls-resources.id}"
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "helm": {
        "source": "helm",
        "version": "2.17.0"
      },
      "kubernetes": {
        "source": "kubernetes",
        "version": "2.38.0"
      },
      "null": {
        "source": "null",
        "version": "3.2.4"
      }
    }
  }
}"
`;
